snuba/request/validation.py

```diff
@@ -10,10 +13,16 @@ def build_request(
     body, schema: RequestSchema, timer: Timer, dataset: Dataset, referrer: str
 ) -> Request:
     with sentry_sdk.start_span(description="build_request", op="validate") as span:
-        request = schema.validate(body, dataset, referrer)
-        span.set_data("snuba_query", request.body)
+        try:
+            request = schema.validate(body, dataset, referrer)
+            span.set_data("snuba_query", request.body)
 
-        timer.mark("validate_schema")
-        # TODO: Refactor to track the invalid query metric.
+            timer.mark("validate_schema")
+            return request
 
-    return request
+        except (InvalidJsonRequestException, InvalidQueryException) as exception:
```

Comment:
I tried to couple the code that logs a successful query with the code that logs an invalid request in a several ways. I think there would be a value in having this logging done in the same place to ensure breaking the invariant that a query is logged only once would be very hard.
This would have looked like a RequestExecutor that handles the pipeline from parsing to execution. Or simply having build_request and parse_and_run_query in the same place.
Unfortunately, the code that builds the request lives here in the request module, while the code that logs successful requests lives in the web module. So no matter how I refactored that code I was introducing new dependencies from internal code to the web module (whether it was here from the request or from the subscriptions/data module).

In the end, I think, we need to move the query execution code out of the web module first before we can couple it with the request builder.